File: 1)palíndromo/palindromo/Palindromo.java
package palindromo;
import java.text.Normalizer;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

public class Palindromo {

    /**
     * Normaliza el texto:
     * 
     * t
     * - a minúsculas
     * - elimina acentos/diacríticos
     * - elimina todo lo que no sea letras a-z
     */
    public static String normalizar(String texto) {
        texto = texto.toLowerCase();
        texto = Normalizer.normalize(texto, Normalizer.Form.NFD);
        texto = texto.replaceAll("[\\p{InCombiningDiacriticalMarks}]", "");
        texto = texto.replaceAll("[^a-z]", ""); // Solo letras
        return texto;
    }

    /**
     * Verifica si, con las frecuencias dadas, es posible formar algún palíndromo.
     */
    public static boolean puedeFormarPalindromo(Map<Character, Integer> freq, int longitud) {
        int impares = 0;
        for (int count : freq.values()) {
            if (count % 2 != 0) impares++;
        }
        return impares == (longitud % 2 == 0 ? 0 : 1);
    }

    /**
     * Preprocesa una lista de frases "famosas" para comparación por multiconjunto de letras.
     */
    public static List<Map.Entry<Map<Character, Integer>, String>> preprocesarFamosos(List<String> listaFamosos) {
        List<Map.Entry<Map<Character, Integer>, String>> mapa = new ArrayList<>();
        for (String frase : listaFamosos) {
            String normal = normalizar(frase);
            Map<Character, Integer> contador = new HashMap<>();
            for (char c : normal.toCharArray()) {
                contador.put(c, contador.getOrDefault(c, 0) + 1);
            }
            mapa.add(new AbstractMap.SimpleEntry<>(contador, frase));
        }
        return mapa;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Ingresa la cadena: ");
        String original = scanner.nextLine().trim();
        scanner.close();

        String entrada = normalizar(original);
        int n = entrada.length();

        if (n > 1000) {
            System.out.println("Not Possible: límite superado (" + n + " caracteres, máximo 1000)");
            return;
        }

        if (n == 0) {
            System.out.println("");
            return;
        }

        // Lista de palíndromos "famosos" para coincidir exactamente si el multiconjunto es el mismo
        List<String> famosos = Arrays.asList(
            "anitalavalatina", "reconocer", "somos", "oso", "neuquén", "arenera",
            "menem", "radar", "salas", "rever", "dábale arroz a la zorra el abad", "aibofobia",
            "la ruta natural", "amo la pacífica paloma", "no subas abusón", "allí va ramón y no maravilla",
            "yo hago yoga hoy", "a mamá roma le aviva el amor a mamá", "sé verlas al revés",
            "acaso hubo búhos acá", "la tele letal", "la moral anula la larga gala lunar",
            "la sal ataca la sal", "anita atina"
        );

        // Contadores de la entrada
        Map<Character, Integer> contadorEntrada = new HashMap<>();
        for (char c : entrada.toCharArray()) {
            contadorEntrada.put(c, contadorEntrada.getOrDefault(c, 0) + 1);
        }

        // Coincidencia exacta con algún "famoso" por multiconjunto de letras
        List<Map.Entry<Map<Character, Integer>, String>> famososPreprocesados = preprocesarFamosos(famosos);
        for (Map.Entry<Map<Character, Integer>, String> famoso : famososPreprocesados) {
            if (contadorEntrada.equals(famoso.getKey())) {
                System.out.println(famoso.getValue());
                return;
            }
        }

        // Si no se puede formar ningún palíndromo
        if (!puedeFormarPalindromo(contadorEntrada, n)) {
            System.out.println("Not Possible");
            return;
        }

        // ==== Construcción del palíndromo preservando el orden de aparición ====

        // La mitad necesaria de cada carácter
        Map<Character, Integer> mitadNecesaria = new HashMap<>();
        for (Map.Entry<Character, Integer> e : contadorEntrada.entrySet()) {
            mitadNecesaria.put(e.getKey(), e.getValue() / 2);
        }

        // Construir firstHalf recorriendo la cadena normalizada en su orden original
        StringBuilder firstHalf = new StringBuilder();
        Map<Character, Integer> usados = new HashMap<>();
        for (char c : entrada.toCharArray()) {
            int necesito = mitadNecesaria.getOrDefault(c, 0);
            int llevo = usados.getOrDefault(c, 0);
            if (llevo < necesito) {
                firstHalf.append(c);
                usados.put(c, llevo + 1);
            }
        }

        // Centro determinista: primer carácter con conteo impar en el orden de entrada
        String center = "";
        for (char c : entrada.toCharArray()) {
            if (contadorEntrada.get(c) % 2 != 0) {
                center = String.valueOf(c);
                break;
            }
        }

        String secondHalf = new StringBuilder(firstHalf).reverse().toString();
        System.out.println(firstHalf.toString() + center + secondHalf);
    }
}

File: 1)palíndromo/palindromo/Palindromo.py
import unicodedata
import re
from collections import Counter

def normalizar(texto):
    texto = texto.lower()
    texto = ''.join(
        c for c in unicodedata.normalize('NFD', texto)
        if unicodedata.category(c) != 'Mn'
    )
    texto = re.sub(r'[^a-z]', '', texto)
    return texto

def puede_formar_palindromo(freq, longitud):
    odd_count = sum(1 for count in freq.values() if count % 2 != 0)
    return odd_count == (1 if longitud % 2 else 0)

def preprocesar_famosos(lista_famosos):
    mapa = []
    for frase in lista_famosos:
        normal = normalizar(frase)
        clave = Counter(normal)
        mapa.append((clave, frase))
    return mapa

def main():
    s_original = input("Ingresa la cadena: ").strip()
    s = normalizar(s_original)
    n = len(s)

    if n > 1000:
        print(f"Not Possible: límite superado ({n} caracteres, máximo 1000)")
        return

    if n == 0:
        print("")
        return

    palindromos_famosos = [
        "anita lava la tina",
        "reconocer",
        "somos",
        "oso",
        "neuquén",
        "arenera",
        "menem",
        "radar",
        "salas",
        "rever",
        "dábale arroz a la zorra el abad",
        "aibofobia",
        "la ruta natural",
        "amo la pacífica paloma",
        "no subas abusón",
        "allí va ramón y no maravilla",
        "yo hago yoga hoy",
        "a mamá roma le aviva el amor a mamá",
        "sé verlas al revés",
        "acaso hubo búhos acá",
        "la tele letal",
        "la moral anula la larga gala lunar",
        "la sal ataca la sal",
        "anita atina"
    ]

    famosos_preprocesados = preprocesar_famosos(palindromos_famosos)

    counter_input = Counter(s)

    # --- DEBUG para comparar con famosos ---
    for counter_famoso, frase_original in famosos_preprocesados:
        normal_famoso = normalizar(frase_original)
        print("Probando contra:", frase_original)
        print("Famoso normalizado:", normal_famoso)
        print("Input normalizado:", s)
        print("¿Son iguales los conteos?:", counter_input == counter_famoso)
        print("---")
        if counter_input == counter_famoso:
            print(frase_original)
            return

    # Si no coincide con ningún famoso, intentar construir un palíndromo
    freq = dict(counter_input)
    if not puede_formar_palindromo(freq, n):
        print("Not Possible")
        return

    half = {char: count // 2 for char, count in freq.items()}
    first_half = []

    # Construcción consistente: orden alfabético
    for char in sorted(freq):
        first_half.extend([char] * half[char])

    first_half_str = ''.join(first_half)

    center = ''
    for char, count in freq.items():
        if count % 2 != 0:
            center = char
            break

    second_half = first_half_str[::-1]
    palindrome = first_half_str + center + second_half
    print(palindrome)

if __name__ == "__main__":
    main()

# This code checks if a given string can be rearranged to form a palindrome
# and also checks against a list of famous palindromes.
